internal/pkg/cli/deploy/env.go

```diff
@@ -142,14 +179,213 @@ func NewEnvDeployer(in *NewEnvDeployerInput) (*envDeployer, error) {
 func (d *envDeployer) Validate(mft *manifest.Environment) error {
 	return d.validateCDN(mft)
 }
 
+// UploadEnvArtifactsOutput is the output of UploadArtifacts.
+type UploadEnvArtifactsOutput struct {
+	AddonsURL          string
+	CustomResourceURLs map[string]string
+}
+
+// UploadArtifacts uploads the deployment artifacts for the environment.
+func (d *envDeployer) UploadArtifacts() (*UploadEnvArtifactsOutput, error) {
+	resources, err := d.getAppRegionalResources()
+	if err != nil {
+		return nil, err
+	}
+	if err := d.patcher.EnsureManagerRoleIsAllowedToUpload(resources.S3Bucket); err != nil {
+		return nil, fmt.Errorf("ensure env manager role has permissions to upload: %w", err)
+	}
+	addonsURL, err := d.uploadAddons(resources.S3Bucket)
+	if err != nil {
+		return nil, err
+	}
+	customResourceURLs, err := d.uploadCustomResources(resources.S3Bucket)
+	if err != nil {
+		return nil, err
+	}
+	return &UploadEnvArtifactsOutput{
+		AddonsURL:          addonsURL,
+		CustomResourceURLs: customResourceURLs,
+	}, nil
+}
+
+// AddonsTemplate returns the environment addons template.
+func (d *envDeployer) AddonsTemplate() (string, error) {
+	if d.addons.notFound {
+		return "", nil
+	}
+	return d.addons.Template()
+}
+
+// DeployEnvironmentInput contains information used to deploy the environment.
+type DeployEnvironmentInput struct {
+	RootUserARN         string
+	AddonsURL           string
+	CustomResourcesURLs map[string]string
+	Manifest            *manifest.Environment
+	ForceNewUpdate      bool
+	RawManifest         []byte
+	PermissionsBoundary string
+	DisableRollback     bool
+}
+
+// GenerateCloudFormationTemplate returns the environment stack's template and parameter configuration.
+func (d *envDeployer) GenerateCloudFormationTemplate(in *DeployEnvironmentInput) (*GenerateCloudFormationTemplateOutput, error) {
+	stackInput, err := d.buildStackInput(in)
+	if err != nil {
+		return nil, err
+	}
+	oldParams, err := d.envDeployer.DeployedEnvironmentParameters(d.app.Name, d.env.Name)
+	if err != nil {
+		return nil, fmt.Errorf("describe environment stack parameters: %w", err)
+	}
+	lastForceUpdateID, err := d.envDeployer.ForceUpdateOutputID(d.app.Name, d.env.Name)
+	if err != nil {
+		return nil, fmt.Errorf("retrieve environment stack force update ID: %w", err)
+	}
+	stack := d.newStackSerializer(stackInput, lastForceUpdateID, oldParams)
+	tpl, err := stack.Template()
+	if err != nil {
+		return nil, fmt.Errorf("generate stack template: %w", err)
+	}
+	params, err := stack.SerializedParameters()
+	if err != nil {
+		return nil, fmt.Errorf("generate stack template parameters: %w", err)
+	}
+	return &GenerateCloudFormationTemplateOutput{
+		Template:   tpl,
+		Parameters: params,
+	}, nil
+}
+
+// DeployEnvironment deploys an environment using CloudFormation.
+func (d *envDeployer) DeployEnvironment(in *DeployEnvironmentInput) error {
```

Comment:
Is there any difference here? (just asking to make sure I'm not overlooking something)